Sgress454/23243 add pagination to manage automations

sgress454 opened the pull request:
# Checklist for submitter

<!-- Note that API documentation changes are now addressed by the product design team. -->

- [ ] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/Committing-Changes.md#changes-files) for more information.
- [ ] Added/updated automated tests
- [ ] A detailed QA plan exists on the associated ticket (if it isn't there, work with the product group's QA engineer to add it)
- [ ] Manual QA for all new/changed functionality

## Details

This PR updates the policy Manage Automations modals to support pagination.  Previously, these modals received a list of policies from the main Manage Policies page, which is itself paginated, so that a user could only add automations to whatever policies were currently listed on the Manage Automations page.  This PR does some refactoring via the creation of a new PaginatedList component which:

* accepts a `fetchPage` property it can call to get a page of data, 
* renders the data in a list with checkboxes and optional custom markup (e.g. dropdowns)
* keeps track of changed ("dirty") items in the list, even across page changes
* allows parent components to access the list of dirty items via a React `ref`

For this specific use case, there's also a new `PoliciesPaginatedList` which implements the `fetchPage` for getting a page of policies, and adds Save and Cancel buttons.  Each of the updated modals uses `PoliciesPaginatedList` to replace its current code for rendering policies in a list, and delegates much of the logic around change tracking to the new components.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/CalendarEventsModal/CalendarEventsModal.tsx:
```diff
@@ -133,28 +111,14 @@ const CalendarEventsModal = ({
       setFormErrors(errors);
-    } else {
-      onSubmit(formData);
+    } else if (paginatedListRef.current) {
+      const changedPolicies = paginatedListRef.current.getDirtyItems();
+      onSubmit({ ...formData, changedPolicies });
     }
```
The `PoliciesPaginatedList` component sends the list of changed items when it calls `onSubmit()`, but since this method (and others like it in the other modals) is also called when the enter key is pressed in the modal, we need another way to get the set of items, hence the call to `getDirtyItems()`.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/CalendarEventsModal/CalendarEventsModal.tsx:
```diff
@@ -133,28 +111,14 @@ const CalendarEventsModal = ({
     setFormData(newFormData);
   };
 
-  const onPolicyEnabledChange = useCallback(
-    (newVal: { name: FormNames; value: boolean }) => {
-      const { name, value } = newVal;
-      const newFormPolicies = formData.policies.map((formPolicy) => {
-        if (formPolicy.name === name) {
-          return { ...formPolicy, isChecked: value };
-        }
-        return formPolicy;
-      });
-      const newFormData = { ...formData, policies: newFormPolicies };
-      setFormData(newFormData);
-    },
-    [formData]
-  );
```
This is now handled by the `PoliciesPaginatedList` and `PaginatedList` components.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/CalendarEventsModal/CalendarEventsModal.tsx:
```diff
@@ -396,7 +323,15 @@ const CalendarEventsModal = ({
-      {configured ? renderConfiguredModal() : renderPlaceholderModal()}
+      <>
+        {configured ? renderConfiguredModal() : renderPlaceholderModal()}
+        {showPreviewCalendarEvent && (
+          <CalendarEventPreviewModal
+            onCancel={togglePreviewCalendarEvent}
+            policy={selectedPolicyToPreview}
+          />
+        )}
```
Moved the calendar preview to be a nested modal inside CalendarEventsModal so that the PoliciesPaginatedList can maintain its state (otherwise it loses the page its on whenever the preview is opened).

jacobshandling replied:
We've generally been avoiding nested modals as a product rule. Would it be feasible to include the
lost state in `PaginatedList`'s `useImperativeHandle`, and have the parent grab that state to
persist before rendering the preview modal, then pass it back in when it goes back to the paginated
list view?

If not, maybe just cosmetically hide the evidence of the parent modal when rendering the child? It
currently looks like:

jacobshandling replied:
<img width="668" alt="Screenshot 2025-02-18 at 11 30 04 AM" src="https://github.com/user-attachments/assets/ba0fecc6-6f47-4840-b8f4-fa14630a4955" />

sgress454 replied:
Handled this by hiding the overlay for the first modal when the second is opened.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/InstallSoftwareModal/InstallSoftwareModal.tsx:
```diff
@@ -42,13 +40,6 @@ interface ISwDropdownField {
   name: string;
   value: number;
 }
-interface IFormPolicy {
-  name: string;
-  id: number;
-  installSoftwareEnabled: boolean;
-  swIdToInstall?: number;
-  platform: CommaSeparatedPlatformString;
-}
```
This is now exported from PoliciesPaginatedList and shared with all the modals.

sgress454 commented on frontend/components/forms/fields/Checkbox/Checkbox.tsx:
```diff
@@ -118,7 +118,7 @@ const Checkbox = (props: ICheckboxProps) => {
           ref={inputRef}
           name={name}
-          checked={value || undefined}
+          checked={!!value}
```
Without this change I get a warning from React about moving from an uncontrolled to a controlled input when the checkbox was checked, which makes sense since we're going from `undefined` to `true`.  I've been unable to figure out why we were getting this warning previously though.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/ManagePoliciesPage.tsx:
```diff
@@ -538,35 +538,11 @@ const ManagePolicyPage = ({
   ) => {
     try {
       setIsUpdatingPolicies(true);
-      const changedPolicies = formData.filter((formPolicy) => {
-        const prevPolicyState = policiesAvailableToAutomate.find(
-          (policy) => policy.id === formPolicy.id
-        );
-
-        const turnedOff =
-          prevPolicyState?.install_software !== undefined &&
-          formPolicy.installSoftwareEnabled === false;
-
-        const turnedOn =
-          prevPolicyState?.install_software === undefined &&
-          formPolicy.installSoftwareEnabled === true;
-
-        const updatedSwId =
-          prevPolicyState?.install_software?.software_title_id !== undefined &&
-          formPolicy.swIdToInstall !==
-            prevPolicyState?.install_software?.software_title_id;
-
-        return turnedOff || turnedOn || updatedSwId;
-      });
-      if (!changedPolicies.length) {
-        renderFlash("success", "No changes detected.");
-        return;
-      }
```
This is all handled internally by the PaginatedList component and the modals.

jacobshandling commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,189 @@
+}
+
+// Wrap with forwardRef to expose the imperative handle:
+const PaginatedList = forwardRef(PaginatedListInner) as <TItem>(
```
It looks like `forwardRef` will be [deprecated soon](https://react.dev/reference/react/forwardRef),
is there a way to do this with the `ref` passed in as a prop instead?

jacobshandling replied:
Ah I see...until we're on v19, you need to use `forwardRef` for the `useImperativeHandle` above. Can
we just leave a note then to change this whenever we upgrade to v19?

jacobshandling commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,189 @@
+    },
+  }));
+
+  // TODO -- better error state?
```
`DataError` 👍

jacobshandling commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,189 @@
+  useEffect(() => {
+    let isCancelled = false;
+
+    async function loadPage() {
```
Would it make sense to move this definition outside of the effect?

sgress454 replied:
Not in this case, because it's asynchronous, and `useEffect` callbacks can't be asynchronous.  Also it uses `isCancelled` as a closure variable.

jacobshandling commented:
Great work! A couple more things

Looks like we lost name truncation on the calendar events modal:
<img width="609" alt="Screenshot 2025-02-18 at 11 27 32 AM" src="https://github.com/user-attachments/assets/292e3acd-4a9c-45f3-95a4-3633852fd7e2" />

We lost a bit of form validation in the RunScript and InstallSoftware modals - in each of them, if any policies are selected but no script/software respectively is chosen for the selected policy, we want to disable the submit button:

<img width="634" alt="Screenshot 2025-02-18 at 11 28 20 AM" src="https://github.com/user-attachments/assets/4199f765-9d5a-4741-8f76-80250eb234d5" />
<img width="634" alt="Screenshot 2025-02-18 at 11 28 28 AM" src="https://github.com/user-attachments/assets/ac65e514-276f-41d8-bfd8-e28e7ada4977" />

sgress454 commented:
> We lost a bit of form validation in the RunScript and InstallSoftware modals - in each of them, if any policies are selected but no script/software respectively is chosen for the selected policy, we want to disable the Save button:

I have an [ongoing discussion on this](https://github.com/fleetdm/fleet/issues/23243#issuecomment-2657424566) because I'm not sure it's a good idea when paginating, but I'll implement it as designed in the meantime.

> Looks like we lost name truncation on the calendar events modal:

Good catch, meant to wrap the labels in PaginatedList in `TooltipTruncatedText`, that should take care of all of them.

sgress454 commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,189 @@
+  // The size of the page to fetch and show.
+  pageSize?: number;
+  // The total # of items on all pages.
+  totalItems: number;
```
That's TODO but I'd really like to update the Pagination control to accept it, because currently it just disables `nextPage` if the current # of page items is <= the size of a page, which means that if the last page has exactly `numItemsOnPage` items, we let you click "next" again and get an empty page:
<img width="806" alt="image" src="https://github.com/user-attachments/assets/94265bc8-cfee-404a-8cd6-56307bd0eabb" />

sgress454 replied:
Actually I think I'll take this out for now, it's definitely worthwhile to do but this PR has got enough in it.

sgress454 commented:
@jacobshandling updated this to:
* use TruncatedTooltipText for the list labels:
<img width="756" alt="image" src="https://github.com/user-attachments/assets/4bec54e8-6308-4f36-9673-e7fe821391b4" />
<img width="737" alt="image" src="https://github.com/user-attachments/assets/8c821390-b466-4594-b9a6-f6122113764a" />

---

* Hide the main calendar events modal when the calendar preview modal is open:
<img src="https://github.com/user-attachments/assets/23856e38-dfe8-4610-9625-3c848232fef6">

--

* Disable the save button on install software and run script modals when dropdown values aren't selected:

![26414-disable-save-1](https://github.com/user-attachments/assets/1535c878-4ee0-410a-95b6-d2689b715165)

---

![26414-disable-save-2](https://github.com/user-attachments/assets/215254a1-0bb8-4df2-9fd5-5970ebd20a92)

---

* Removed the `total` property for now
* Added comment re: using `forwardRef` after React 19

jacobshandling commented:
Looks like we need to make the background colors line up for the nested modals:
![ezgif-8a7bf8b3d47142](https://github.com/user-attachments/assets/7b54ef45-c8a9-498b-8757-276a071ca113)

jacobshandling commented on frontend/pages/policies/ManagePoliciesPage/components/CalendarEventsModal/_styles.scss:
```diff
@@ -36,10 +36,30 @@
     margin-top: $pad-large;
   }
+
+  &__hide-main {
```
Nice. I think you'll need to hide the app-wide overlay from the original modal as well, or something
like that.

sgress454 replied:
Done.

jacobshandling commented on frontend/pages/policies/ManagePoliciesPage/components/PoliciesPaginatedList/PoliciesPaginatedList.tsx:
```diff
@@ -0,0 +1,198 @@
+            orderDirection: "asc" as const,
+            orderKey: DEFAULT_SORT_COLUMN,
+            teamId,
+            mergeInherited: false,
```
Do we want to set this conditionally to allow merging inherited policies for the Other workflows modal?

jacobshandling commented on frontend/pages/policies/ManagePoliciesPage/components/PoliciesPaginatedList/PoliciesPaginatedList.tsx:
```diff
@@ -0,0 +1,198 @@
+  const fetchPage = useCallback((pageNumber: number) => {
+    let fetchPromise;
+
+    if (teamId === APP_CONTEXT_ALL_TEAMS_ID) {
```
Very cool to see how we can interact a bit lower down with `useQueryClient`! Have you checked out how we [conditionally fetch global or team queries](https://github.com/fleetdm/fleet/blob/b990b3c6d92092854a709060b74a16f14b5133cd/frontend/pages/queries/ManageQueriesPage/ManageQueriesPage.tsx#L123-L155)? By setting the query key's `teamId` and `mergeInherited` fields based on `teamIdForApi`, we get the desired API calls. I think this would be a bit cleaner here, unless there's some specific reason to use `useQueryClient`?

sgress454 replied:
The queries API handles team vs. global queries on the back end, but alas the policies API has separate endpoints for team vs. global.  As for using `useQueryClient()`, if you mean as opposed to `useQuery()`, this is so that we can return a promise to the lower-level `PaginatedList` component, rather than doing the query automatically here in the PoliciesPaginatedList component.  The idea of `PaginatedList` is for it to contain all of the data and state itself, rather than the parent having to keep track of the data and what page we're on.

jacobshandling commented on frontend/pages/policies/ManagePoliciesPage/components/PoliciesPaginatedList/PoliciesPaginatedList.tsx:
```diff
@@ -0,0 +1,198 @@
+            orderDirection: "asc" as const,
+            orderKey: DEFAULT_SORT_COLUMN,
+            teamId,
+            mergeInherited: false,
```
Maybe make this conditional to allow merging inherited for the Other workflows modal, which seems to work for global policies?

sgress454 replied:
It's sneaky but the inherited (global) policies are currently filtered out of the main list before being sent to the modals [here](https://github.com/fleetdm/fleet/blob/main/frontend/pages/policies/ManagePoliciesPage/ManagePoliciesPage.tsx#L310-L313), so for our use case we can always do `mergeInherited: false`.

jacobshandling reviewed (commented):
Great stuff, I'm taking notes on how to build powerful, flexible abstractions. Just the couple suggestions.

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/CalendarEventPreviewModal/_styles.scss:
```diff
@@ -78,3 +78,9 @@
   }
 }
+
+// Pop up preview immediately.
+.modal__background:has(.calendar-event-preview-modal) {
+  transition: none;
+  animation: none;
```
Without this, there's a background flicker as the first modal's background disappears immediately and the preview modal's background animates.

RachelElysia commented on frontend/pages/policies/ManagePoliciesPage/ManagePoliciesPage.tsx:
```diff
@@ -1055,6 +1007,8 @@ const ManagePolicyPage = ({
             isUpdating={isUpdatingPolicies}
             onExit={toggleOtherWorkflowsModal}
             onSubmit={onUpdateOtherWorkflows}
+            teamId={currentTeamId ?? 0}
```
Second [Figma spec](https://www.figma.com/design/7Oeb9nPQF51Fgwiik2aeJt/%2323243-Add-pagination-to-automations-modal-on-Policies-and-disable-button-if-not-available?node-id=5461-3389&t=n2EQRpXUVgYgN4sZ-0) skipped for disabling this

RachelElysia replied:
Third [Figma spec](https://www.figma.com/design/7Oeb9nPQF51Fgwiik2aeJt/%2323243-Add-pagination-to-automations-modal-on-Policies-and-disable-button-if-not-available?node-id=5461-3627&t=n2EQRpXUVgYgN4sZ-0)

RachelElysia replied:
Similar [Figma spec](https://www.figma.com/design/7Oeb9nPQF51Fgwiik2aeJt/%2323243-Add-pagination-to-automations-modal-on-Policies-and-disable-button-if-not-available?node-id=5331-4200&t=n2EQRpXUVgYgN4sZ-0) skipped on Software Page

RachelElysia replied:
> Similar [Figma spec](https://www.figma.com/design/7Oeb9nPQF51Fgwiik2aeJt/%2323243-Add-pagination-to-automations-modal-on-Policies-and-disable-button-if-not-available?node-id=5331-4200&t=n2EQRpXUVgYgN4sZ-0) skipped on Software Page

@sgress454 I'm still not seeing this in your branch, I even VS Code searched "to manage" and its not popping up

RachelElysia replied:
@marko-lisica - I noticed this ticket's Figma files has created a new pattern for Manage automations dropdown buttons disabled with tooltips for policies and software, but it doesn't include the new pattern for manage automations queries (which still just hides it instead of what I would expect is to also follow the new pattern of disabling it with a tooltip). Was that intentional.

sgress454 replied:
Pushed up the software page change!

marko-lisica replied:
> @marko-lisica (product designer on this) - I noticed this ticket's Figma files has created a new pattern for Manage automations dropdown buttons disabled with tooltips for policies and software, but it doesn't include the new pattern for manage automations queries (which still just hides it instead of what I would expect is to also follow the new pattern of disabling it with a tooltip). Was that intentional?

@RachelElysia It isn't intentional. I probably missed that. I assume we hide "Manage automation" if there's no queries on the team?

RachelElysia reviewed (changes requested):
Hola! Code looks clean af! Just a lot of UI discrepancies between Figma designs and testing this branch. Feel free to assign me to re-review when the comments are addressed!

sgress454 commented on frontend/pages/policies/ManagePoliciesPage/components/PoliciesPaginatedList/PoliciesPaginatedList.tsx:
```diff
@@ -0,0 +1,218 @@
+  return (
+    <div className={`${baseClass} form`}>
+      <div className="form-field">
+        <div className={labelClasses}>Policies:</div>
```
Fixed!

sgress454 commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,211 @@
+          const item = dirtyItems[_item[idKey]] ?? _item;
+          return (
+            <li className={`${baseClass}__row`} key={item[idKey]}>
+              <Checkbox
```
It's even weirder than than that -- a narrow band is clickable across the whole row (i.e. the checkbox and its label), but not the padding around that.  Fixed now to make the whole row clickable.

sgress454 reviewed (commented):
I made some updates and I think addressed all of your comments, let me know!

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,211 @@
+                {renderItemLabel ? (
+                  renderItemLabel(item)
+                ) : (
+                  <TooltipTruncatedText
```
I'm still seeing this, I think it's happening when it doesn't overflow the full width, but overflows past the dropdown?

It also looks better with:
```
.tooltip-truncated-text .tooltip-truncated {
    max-width: 101%;
    }
```
instead of `100%` and
```
.tooltip-truncated-text .truncated {
max-width: 100%;
}
```
instead of `99%`;
because it is cutting off the  end of the string with `"(No team)"` and putting `"(No team..."`

But that may have reprocussions elsewhere in the app >.<


Uploading Screen Recording 2025-02-27 at 9.38.30 AM.mov…

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,225 @@
+import Spinner from "components/Spinner";
+import TooltipTruncatedText from "components/TooltipTruncatedText";
+// @ts-ignore
+import Pagination from "components/Pagination";
```
Don't hate the messenger >.<

Just noticed this on `components/Pagination`:

```
/**
 * WARNING: DEPRICATED:
 * This pagination component is DEPRICATED. It is being kept around until we replace its
 * use. For now use the Pagination component in the pages/ManageControlsPage/components.
 */
```

Is there a way we can move to the other pagination?

sgress454 replied:
Ok, looking into it.

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,211 @@
+        resultsOnCurrentPage={items.length}
+        currentPage={currentPage}
+        resultsPerPage={pageSize}
+        onPaginationChange={setCurrentPage}
```
> There is, for some reason I had the page size set to 8. Fixed to 10. Incidentally we should fix the Pagination component to accept a total so it can render the last page correctly if it has exactly pageSize elements on it...

I looked it up because I vaguely remembered we started solving this problem... 
We started adding meta data to our API responses, which would require a API call without inherited policies:
``` 
  meta: {
    has_next_results: true,
    has_previous_results: false,
  },
```

Since this is a bug that will only be seen now that we have pagination, we should file it as an unreleased bug and see what we can do to fix it before release?

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,279 @@
+                ) : (
+                  <TooltipTruncatedText
+                    value={
+                      <span className={`${baseClass}__item-label`}>
```
Just tested, looks like we can simplify the code to just ` <TooltipTruncatedText value={item[labelKey]} />`

sgress454 replied:
Done! I may have been thinking about parent components styling the label using that class, but they can provide a label styling function if they want.

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,279 @@
+import Checkbox from "components/forms/fields/Checkbox";
+import Spinner from "components/Spinner";
+import TooltipTruncatedText from "components/TooltipTruncatedText";
+// @ts-ignore
```
No longer need the `// @ts-ignore` since the new Pagination is typescript file <3

RachelElysia commented on frontend/components/PaginatedList/PaginatedList.tsx:
```diff
@@ -0,0 +1,279 @@
+    },
+  }));
+
+  function disableNext() {
```
Just heads up our general pattern for functions is 
`const disableNext = () => {` instead of using `function`

RachelElysia replied:
I don't know if we're moving away from this or not, so let's not change what you got

sgress454 replied:
I'll fix this -- `const` is a better pattern as it doesn't get hoisted.  Will fix.

RachelElysia reviewed (dismissed):
Nothing blocking approval, open for iteration on any unresolved comments
